### ISSUE TITLE
Count unused files

### ISSUE THREAD

**Issue by anderspitman** [codecov/support#292]
This might be a stupid question. It appears that codecov only runs coverage for files that are actually compiled. So lets say I add a new source file to my project, and I compile it for my program binary, but I forget to add unit tests for the new file. Lets say further that I don't have any integration tests for the full program (obviously a bad idea). In this situation, the new file will never be compiled during my unit testing, and therefore no coverage will be generated for that file. It would be desirable for the file to be included in the report with 0% coverage.

Is there a way to get this behavior?

**Comment by stevepeak** [comment]
@anderspitman great question. Codecov simply expresses only coverage data provided. So if the file is not compiled and no coverage reports are generated then Codecov will not include this file. Every language has their own technique to include files that may not be compiled or executed, and making sure they are tracked by coverage.

**Comment by reedacartwright** [comment]
@stevepeak Do you know what the trick is for C++/gcov and the codecov-bash?

**Comment by stevepeak** [comment]
@reedacartwright `gcov` does require `.gcno` or `.cc` files to exist to execute gcov against. Do you know if these types of files exist for the missing files? If so, perhaps they were not discovered by the bash uploader.

**Comment by anderspitman** [comment]
@stevepeak from my limited empirical testing, it appears even if the `.gcno` files are present, nothing is reported by codecov if the `.gcda` files are not also present. Is that correct? Would it make sense to perhaps detect if the `.gcno` (generated at compile time) exists for a file and report 0% coverage if there is not a corresponding `.gcda` (generated at execution time) file?

**Comment by stevepeak** [comment]
> it appears even if the .gcno files are present, nothing is reported by codecov if the .gcda files are not also present

Correct. If the `.gcno`'s are found and processed (which they should be) then you get report data (even if empty). No `.gnco` no report

> Would it make sense to perhaps detect if the .gcno (generated at compile time) exists for a file and report 0% coverage if there is not a corresponding .gcda (generated at execution time) file?

This is an interesting idea and it's very possible. I won't have a chance to work on this for a while, and would welcome a pull request to codecov/codecov-bash. Are you familiar with bash?

**Comment by anderspitman** [comment]
@stevepeak only the minimum necessary to be proficient. Not enough for something like this.

**Comment by GDYendell** [comment]
I needed them same thing and found this. I have used the following to make this work for me:

`find . -name "*.gcno" -exec sh -c 'touch -a "${1%.gcno}.gcda"' _ {} \;`

I do this between executing my tests and running gcov on the resulting .gcda files. All it does is search for any .gcno files without a matching .gcda file (meaning the file hasn't been executed) and creates a dummy .gcda file. gcov then reads those files and finds 0% coverage. I would appreciate hearing any improvements on this method, but it works as far a I can tell.

Coverage report is [here](https://codecov.io/gh/GDYendell/odin-data/list/386c032ae0e9fa6583d11334c7883964a45ba9f1). As you can see there are lots of files with 0% coverage.

**Comment by stevepeak** [comment]
@GDYendell awesome find! Perhaps we should add this to our bash uploader as a default. Thoughts?

**Comment by jmcker** [comment]
@stevepeak @GDYendell The method mentioned here no longer works with gcov v5.4.

gcov ignores the empty .gcda file with something like:
`./CMakeFiles/simple-math.dir/src/simplemath/src/SimpleMath.cpp.gcda:not a gcov data file`

I worked around this using `lcov` [[here](https://github.com/symboxtra/SplitSound-Desktop/blob/c93e3c118327268b9fa65bfe7f361e3c1c8e0b96/test.sh)]. Per the [man page](http://ltp.sourceforge.net/coverage/lcov/lcov.1.php):

```
	-i
	--initial
		  Capture initial zero coverage data.

		  Run  lcov  with -c and this option on the directories containing
		  .bb, .bbg or .gcno files  before  running  any  test  case.  The
		  result  is  a  "baseline"  coverage data file that contains zero
		  coverage for every instrumented line.  Combine  this  data  file
		  (using  lcov  -a) with coverage data files captured after a test
		  run to ensure that the percentage of total lines covered is cor‐
		  rect  even when not all source code files were loaded during the
		  test.

		  Recommended procedure when capturing data for a test case:

		  1. create baseline coverage data file
				 # lcov -c -i -d appdir -o app_base.info

		  2. perform test
				 # appdir/test

		  3. create test coverage data file
				 # lcov -c -d appdir -o app_test.info

		  4. combine baseline and test coverage data
				 #   lcov   -a   app_base.info   -a    app_test.info    -o
				 app_total.info
```

Following this procedure and turning off CodeCov's gcov option using `-X gcov` worked for me.
I think CodeCov will handle merging the base and test reports on its own already, removing the need for step 4.

Since `lcov` isn't readily installed on many machines, I don't know if it would be viable to add the option to the bash uploader (perhaps opt-in instead of gcov's opt-out?). If you think it would be beneficial to enough people though, I would be happy to spend some time on it.

**Comment by GDYendell** [comment]
@jmcker @stevepeak This seems like the way it _should_ be done!
